packages/react-flex/Spec.md

```diff
@@ -0,0 +1,540 @@
+| Additional CSS class name(s) to apply                                                    | `className`            | `className`                  | matching         |
+| Additional CSS styles to apply to the component instance                                 | extends `HTMLElement`  | `styles`                     | matching         |
+| Override for theme site variables to allow modifications of component styling via themes | through `IStackTokens` | `variables`                  | different naming |
+| Defines a custom order for the item                                                      | `order`                | -                            | Flex missing     |
+| Defines whether the item should be prevented from shrinking                              | `disableShrink`        | -                            | Flex missing     |
+| Defines whether the item should take up 100% of the height of its parent                 | `verticalFill`         | -                            | Flex missing     |
+| Push item towards opposite side in the container's direction                             | -                      | `push`                       | Stack missing    |
+| Defines the flex-basis for the item                                                      | -                      | `size`                       | Stack missing    |
+| -                                                                                        | -                      | `flexDirection` [DEPRECATED] | -                |
+| -                                                                                        | -                      | `design`                     | -                |
+
+## API proposal
+
+<!-- link to the component's `.types.ts` file -->
+
+### Flex
+
+| Name              | Type                                                                                                  | Default value | Comments                                                                                                                                                                                                                                 |
+| ----------------- | ----------------------------------------------------------------------------------------------------- | ------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `direction`       | string = "`row`", "`row-reverse`", "`column`", "`column-reverse`"                                     | "`row`"       | To further cement the idea of Flex being an abstraction of flexbox, this would bring the user to a closer experience of flexbox while also being less biased on our interpretation of the default direction. Absorbs Stack's `reversed`. |
```

Comment:
To support easier readability and runtime configuration for applications would it make sense to break this into something like:
* layout -> 'row' | 'column'
* reverse -> boolean

An alternative which seems interesting would be all booleans:
* row -> boolean
* column -> boolean
* reverse -> boolean
However this raises the problem of what happens if `row` and `column` are used at the same time

With the above 2 approaches in an app you can have state that controls all the combinations in a more decoupled way.... for example I want `row` but in an RTL scenario simply add `reverse`

Reply:
I think that pattern would be more cumbersome for our users if we consider most are already aware of how `flex` works. Even if they don't, our components would educate them into how to use it instead of exposing a whole new interface that would be completely useless to them in a different context.
Plus, like you mentioned for the second alternative, we start going into overriding stuff or preparing for unexpected user usage, like we currently have our components. In my opinion, the leaner and closer to `flexbox` the implementation is, the more intuitive and less problematic it would be.
What would be some advantages from this implementation?

Reply:
My main gripe with following flex to the letter here is that as an app developer I'd just end up creating the same abstraction over the string-based API anyway in state, since they are decoupled properties. I was thinking of a use case like this:

```jsx
<Flex row reverse={isRtl} />

// As opposed to some complex conditional spreading
<Flex direction={isRtl ? 'row' : 'row-reverse'}
<Flex direction={isRtl ? isVertical ? .....}
```

Reply:
> I'd just end up creating the same abstraction over the string-based API anyway in state

That should be probably extracted to one common place (a pure function) to mitigate repetition. anyone during code review will probably point that out.

> regarding RTL

If you're developing an app that needs to be served to RTL audience, it should be done on global level, baked in within whole framework to do rendering just right and consistent. By leaving this out to developer, things can go wrong rather quickly.

Adding another layer of artificial API seems overall like a wrong approach
- material ui got this abstraction quite right for example https://material-ui.com/system/flexbox/
- also good inspiration can be taken from angular https://github.com/angular/flex-layout/wiki/Declarative-API-Overview

Reply:
For that specific example I much prefer your implementation. However, I'm thinking of how many cases there would be for on the fly changes to direction and, when existing, if there wouldn't already exist logic behind it regardless. Why not make the end result for that logic the proper `flex-direction` you want your content in?

It seems that for both extremes of the spectrum this would not simplify but rather complicate as the user would need to read the docs for how to use it, instead of simply following the `flexbox` standard.

In my opinion, simply abstracting `flexbox` would be the ideal option not only for us, implementation wise, but also for our users by giving them the freedom of control, intuitive usage (as it's just abstracting `flex-direction`) as well as any new changes to `flexbox` being immediately available to them.

Reply:
> Adding another layer of artificial API seems overall like a wrong approach

So if we want to retain the native flexbox semantics/syntax I actually feel kind of strange now that we are making a `Flex` component that just exposes the standard css properties and I wonder why we don't just make a hook that will apply the correct flexbox styles in a classname ?

> If you're developing an app that needs to be served to RTL audience, it should be done on global level, baked in within whole framework to do rendering just right and consistent. By leaving this out to developer, things can go wrong rather quickly.

Is that an extra requirement that `Flex` should be RTL aware and apply the correct `direction` prop ?

Reply:
> I actually feel kind of strange now that we are making a Flex component 

right, I think @andrefcdias asked this somewhere during his open ui research. I like the idea with hook 👌

Reply:
> Is that an extra requirement that Flex should be RTL aware and apply the correct direction prop 

No idea. I haven't found any docs that describe how fluent ui approaches RTL from design/implementation perspective

Reply:
From what I've seen, there's no docs/requirement for the component to be RTL aware nor any implementation in both of the libs that suggests we do it already.

Also, I like the hook idea and given our styling system, it would work. My only concerns are how we can maintain the same level of visual information about the styling on the component if we're now just applying a class and how the usage would look like. Would this also address the issue at hand? What are your visions for a flex hook usage?

Reply:
The pro of hooks is that we no longer need to worry about compose:

```jsx
<section className={useFlexStyles()} />

<Flex as="section" />
```

> visual information about the styling on the component

I guess you won't see `<Flex />` in your codebase, but tbf that's just CSS in general ?

Reply:
I dunno if it's better, but as I said before creating a react component to write css in jsx seems strange...

Reply:
> What are your visions for a flex hook usage?

From "compatibility" reasons I'd go with your proposal as is. As a "side-effect" I'd encapsulate all logic within a hook which could be exposed as a public API

```
// CONTRIVED EXAMPLE

// @file flex.tsx
import {useFlex} from './flex.hook'

export function Flex(props: Props){
   const {css, root: Root} = useFlex(props)

   return <Root className={css}>{props.children}</Root>
}

// @file index.ts
export * form './flex'
export * from './flex.hook'
```

Reply:
Would `useFlexStyles()` strictly return a class that has `display: flex` in it?
This is what I was picturing for a very simple example and it seems visually very cluttered to me:

```HTML
<section className={useFlexStyles({
    direction: "column"
})}>
  <span>1</span>
  <span className={useFlexItemStyles({
    grow: 1
  })}>2</span>
</section>
```

Reply:
> and it seems visually very cluttered to me

oh right this doesn't feel right.

probably we don't need hook at all ? (there shouldn't be any state, or sideffect needed) - I might miss some implementation info related to how our custom CSS in JS is supposed to work)

Reply:
Hmm yeah it looks kind of strange.... let's just wrap it up here it's not going to go any more productive. I'm fine with religiously following CSS native syntax until we get feedback that tells us otherwise